src/renderers/shared/reconciler/__tests__/ReactCompositeComponent-test.js

```diff
@@ -1241,4 +1241,51 @@ describe('ReactCompositeComponent', function() {
 
   });
 
+  it('should warn when mutated props are passed', function() {
+
+    var container = document.createElement('div');
+
+    function Foo (props) {
+      React.Component.call(this, props);
+    }
+
+    Foo.prototype = Object.create(React.Component.prototype, {
+      constructor: {
+        value: Foo,
+      },
+    });
+
+    Foo.prototype.render = function() {
+      return <span />;
+    };
+
+    function Bar (props) {
```

Comment:
Sure. Foo (or Bar) can directly inherit. Making the change in the next commit.

Is ES6 necessary here? I find most of the React Component definition is ES5 syntax. Simpler as it maybe, mixing up the syntax looks bad on the file as a whole.